ssl/quic/qlog.c

```diff
@@ -0,0 +1,702 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/qlog.h"
+#include "internal/json_enc.h"
+#include "internal/common.h"
+#include "internal/cryptlib.h"
+
+# define BITS_PER_WORD (sizeof(size_t) * 8)
+# define NUM_ENABLED_W ((QLOG_EVENT_TYPE_NUM + BITS_PER_WORD - 1) / BITS_PER_WORD)
+
+static ossl_unused ossl_inline int bit_get(const size_t *p, uint32_t bit_no)
+{
+    return p[bit_no / BITS_PER_WORD] & (((size_t)1) << (bit_no % BITS_PER_WORD));
+}
+
+static ossl_unused ossl_inline void bit_set(size_t *p, uint32_t bit_no, int enable)
+{
+    size_t mask = (((size_t)1) << (bit_no % BITS_PER_WORD));
+
+    if (enable)
+        p[bit_no / BITS_PER_WORD] |= mask;
+    else
+        p[bit_no / BITS_PER_WORD] &= mask;
+}
+
+/* TODO abort */
```

Comment:
???

Reply:
Heh, this isn't about abort() if you're curious &mdash; at one point I was thinking we may need the ability to log entries transactionally and roll back if an error occurs during the process of logging an entry. Currently however the actual logging code is simple enough that this doesn't prove necessary.

Anyway, I forgot to remove this comment; removed.